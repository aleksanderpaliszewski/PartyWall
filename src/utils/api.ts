import axios, {AxiosError, AxiosResponse} from 'axios';
import qs from 'qs';
import urljoin from 'url-join';

import {API_BASE_URL} from '@env';

if (API_BASE_URL) {
  axios.defaults.baseURL = API_BASE_URL;
}

interface APIErrorPayload {
  message: string;
  details?: APIErrorDetails;
}

interface APIErrorDetails {
  [key: string]: string;
}

export class APIError extends Error {
  details: APIErrorDetails;
  status: number;
  code?: number;
  originalError: AxiosError<APIErrorPayload>;

  constructor(error: AxiosError<APIErrorPayload>) {
    if (!error.response || !error.response.data) {
      super(error.message);
      this.details = {};
      this.status = 0;
    } else {
      const {details, message}: APIErrorPayload = error.response!.data;
      super(message || error.message);
      this.details = details || {};
      this.status = error.response!.status;
    }

    this.originalError = error;
  }
}

axios.interceptors.response.use(
  (response) => response,
  (error) => Promise.reject(new APIError(error)),
);

type UnauthorizedHandler = (error: AxiosError) => Promise<never>;

export default class Api {
  handleUnauthorized: UnauthorizedHandler;

  constructor(handleUnauthorized: UnauthorizedHandler) {
    this.handleUnauthorized = handleUnauthorized;
  }

  get<T>(uri: string, query?: object): Promise<AxiosResponse<T>> {
    return axios
      .get<T>(urljoin(uri, stringifyQuery(query)))
      .catch(this.handleUnauthorized);
  }

  post<T>(uri: string, data: object): Promise<AxiosResponse<T>> {
    return axios.post<T>(uri, data).catch(this.handleUnauthorized);
  }

  static setToken(token: string) {
    console.log(token);
    axios.defaults.headers = {
      Authorization: `Bearer ${token}`,
    };
  }

  static resetToken() {
    delete axios.defaults.headers.Authorization;
  }
}

function stringifyQuery(object = {}) {
  return qs.stringify(object, {
    arrayFormat: 'repeat',
    addQueryPrefix: true,
  });
}
