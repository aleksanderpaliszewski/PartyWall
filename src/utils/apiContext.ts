import {createContext} from 'react';
import Api from './api';

const ApiContext = createContext<Api>(new Api((e) => Promise.reject(e)));

export default ApiContext;
