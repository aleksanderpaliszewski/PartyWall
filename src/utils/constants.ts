import {Platform} from 'react-native';

export const isIos = Platform.OS === 'ios';
export const NO_PRODUCTS_TEXT = 'No products in the database...';
