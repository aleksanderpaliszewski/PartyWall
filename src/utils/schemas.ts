import {object, string, number, ref} from 'yup';

export const LogInSchema = object().shape({
  password: string()
    .min(8, 'Password must have a minimum of 8 characters')
    .required('Enter password'),
  username: string().required('Enter username'),
});

export const RegisterSchema = object().shape({
  password: string()
    .min(8, 'Password must have a minimum of 8 characters')
    .required('Enter password'),
  passwordConfirmation: string()
    .oneOf([ref('password')], 'The entered passwords are not the same')
    .required('Repeat password'),
  username: string().required('Enter username'),
});

export const AddProductSchema = object().shape({
  name: string().required('Enter name'),
  price: string().required('Enter price'),
  quantity: number()
    .typeError('Invalid Input: numbers please')
    .required('Enter quantity'),
});
