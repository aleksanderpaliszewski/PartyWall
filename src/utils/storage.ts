import AsyncStorage from '@react-native-community/async-storage';

export default class Storage {
  static setItem(key: string, value: unknown) {
    return AsyncStorage.setItem(key, serialize(value)).catch(console.error);
  }

  static getItem<T>(key: string): Promise<T | null> {
    return AsyncStorage.getItem(key)
      .then((value) => deserialize<T>(value))
      .catch((error) => {
        console.error(error);

        return null;
      });
  }

  static removeItem<T>(key: string): Promise<void> {
    return AsyncStorage.removeItem(key)
      .then(() => console.log(`key: ${key} removed`))
      .catch(console.error);
  }
}

function serialize(value: unknown) {
  return JSON.stringify(value);
}

function deserialize<T>(value: string | null): T {
  return value ? JSON.parse(value) : null;
}
