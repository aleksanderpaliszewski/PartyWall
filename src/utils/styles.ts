import {StatusBar, StyleSheet} from 'react-native';

export const COLORS = {
  primary: '#301cbe',
  secondary: '#730040',
  background: '#F0F0F0',
  border: '#dcdcdc',
  dark: '#322f3d',
  white: '#FFFFFF',
  black: '#000000',
  error: '#AE1C23',
};

export const SPACING = {
  Horizontal: 15,
  Vertical: 10,
};

export default StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    marginHorizontal: 2 * SPACING.Horizontal,
    marginVertical: 2 * SPACING.Vertical,
  },
  grow: {
    flex: 1,
    justifyContent: 'space-between',
  },
  statusBar: {
    marginTop: StatusBar.currentHeight || 0,
  },
  text: {
    color: COLORS.white,
  },
  header: {
    fontSize: 30,
    color: COLORS.white,
  },
  content: {
    fontSize: 16,
  },
  shadow: {
    shadowColor: COLORS.black,
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.1,
    shadowRadius: 2,
    elevation: 2,
  },
  inputContainer: {
    width: '100%',
    borderTopRightRadius: 40,
    borderTopLeftRadius: 40,
    paddingHorizontal: 2 * SPACING.Horizontal,
    paddingTop: 6 * SPACING.Vertical,
    paddingBottom: 3 * SPACING.Vertical,
    backgroundColor: COLORS.white,
  },
  button: {
    marginTop: 2 * SPACING.Vertical,
    alignSelf: 'center',
    borderRadius: 30,
  },
  formContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'flex-end',
  },
});
