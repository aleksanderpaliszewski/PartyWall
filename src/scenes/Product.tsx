import React, {FC, useContext, useState} from 'react';
import {StyleSheet, View} from 'react-native';
import {Button, Switch, Text} from 'react-native-paper';
import {FormikValues} from 'formik';
import {pickBy, identity} from 'lodash';

import Layout from '../components/Layout';
import ProductForm from '../components/ProductForm';
import {HomeStackScreenProps} from '../navigations/Home';
import UserContext from '../utils/userContext';
import SnackBarContext from '../utils/snackbarContext';
import ApiContext from '../utils/apiContext';
import {Scenes} from '../api/enums';

import globalStyles, {COLORS, SPACING} from '../utils/styles';

const Product: FC<HomeStackScreenProps<Scenes.Product>> = ({navigation}) => {
  const api = useContext(ApiContext);
  const {resetAuthData} = useContext(UserContext);
  const {setMessage} = useContext(SnackBarContext);
  const [isDrinkProduct, setIsDrinkProduct] = useState(true);
  const [loading, setLoading] = useState(false);

  const toggleSwitch = () => setIsDrinkProduct(!isDrinkProduct);

  const deleteEmptyFields = (values: FormikValues) => pickBy(values, identity);

  const handleSubmit = (values: FormikValues) => {
    setLoading(true);

    return api
      .post('products', deleteEmptyFields(values))
      .then(() => navigation.replace(Scenes.Home))
      .catch((error) => setMessage(error.message))
      .finally(() => setLoading(false));
  };

  return (
    <Layout>
      <View style={globalStyles.appBar}>
        <Button
          mode="text"
          style={globalStyles.logoutButton}
          labelStyle={globalStyles.logoutButtonLabel}
          onPress={resetAuthData}>
          Log out
        </Button>
      </View>
      <View style={[globalStyles.roundedContainer, styles.container]}>
        <Text style={[globalStyles.header, styles.header]}>Product type</Text>
        <View style={styles.row}>
          <Switch value={!isDrinkProduct} onValueChange={toggleSwitch} />
          <Text style={styles.text}>Food product</Text>
        </View>
        <View style={styles.row}>
          <Switch value={isDrinkProduct} onValueChange={toggleSwitch} />
          <Text style={styles.text}>Drink product</Text>
        </View>
        <Text style={[globalStyles.header, styles.header]}>Product info</Text>
        <ProductForm
          loading={loading}
          handleSubmit={handleSubmit}
          isDrink={isDrinkProduct}
        />
      </View>
    </Layout>
  );
};

const styles = StyleSheet.create({
  container: {
    alignItems: 'flex-start',
    paddingVertical: SPACING.Vertical,
    paddingHorizontal: SPACING.Horizontal,
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start',
    marginVertical: SPACING.Vertical,
    marginHorizontal: SPACING.Horizontal,
  },
  text: {
    marginLeft: SPACING.Horizontal,
  },
  header: {
    color: COLORS.black,
    marginVertical: 2 * SPACING.Vertical,
  },
});

export default Product;
