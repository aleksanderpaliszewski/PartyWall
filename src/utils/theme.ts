import {DefaultTheme, useTheme as useDefaultTheme} from 'react-native-paper';
import {Theme} from 'react-native-paper/src/types';

import {COLORS} from './styles';

const theme: Theme = {
  ...DefaultTheme,
  colors: {
    ...DefaultTheme.colors,
    primary: COLORS.primary,
    accent: COLORS.secondary,
  },
};

export default theme;

export function useTheme() {
  return useDefaultTheme(theme);
}
