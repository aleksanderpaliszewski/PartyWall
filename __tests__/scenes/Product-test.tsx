import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import {shallow, configure} from 'enzyme';

import Product from '../../src/scenes/Product';
import {getProductProps} from '../utils/navigation';

configure({adapter: new Adapter()});
describe('Product', () => {
  it('render', () => {
    const mockProps = getProductProps();

    const wrapper = shallow(<Product {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
