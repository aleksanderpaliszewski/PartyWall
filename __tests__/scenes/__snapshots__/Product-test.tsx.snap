// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product render 1`] = `
<Layout>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "flex-end",
        "paddingHorizontal": 15,
        "paddingVertical": 20,
      }
    }
  >
    <withTheme(Button)
      labelStyle={
        Object {
          "color": "#FFFFFF",
        }
      }
      mode="text"
      onPress={[Function]}
      style={
        Object {
          "alignSelf": "flex-end",
        }
      }
    >
      Log out
    </withTheme(Button)>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderTopLeftRadius": 40,
          "borderTopRightRadius": 40,
          "flex": 10,
          "justifyContent": "flex-start",
          "opacity": 0.9,
          "overflow": "hidden",
        },
        Object {
          "alignItems": "flex-start",
          "paddingHorizontal": 15,
          "paddingVertical": 10,
        },
      ]
    }
  >
    <withTheme(undefined)
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 30,
          },
          Object {
            "color": "#000000",
            "marginVertical": 20,
          },
        ]
      }
    >
      Product type
    </withTheme(undefined)>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginHorizontal": 15,
          "marginVertical": 10,
        }
      }
    >
      <withTheme(Switch)
        onValueChange={[Function]}
        value={false}
      />
      <withTheme(undefined)
        style={
          Object {
            "marginLeft": 15,
          }
        }
      >
        Food product
      </withTheme(undefined)>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-start",
          "marginHorizontal": 15,
          "marginVertical": 10,
        }
      }
    >
      <withTheme(Switch)
        onValueChange={[Function]}
        value={true}
      />
      <withTheme(undefined)
        style={
          Object {
            "marginLeft": 15,
          }
        }
      >
        Drink product
      </withTheme(undefined)>
    </View>
    <withTheme(undefined)
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 30,
          },
          Object {
            "color": "#000000",
            "marginVertical": 20,
          },
        ]
      }
    >
      Product info
    </withTheme(undefined)>
    <AddProductForm
      handleSubmit={[Function]}
      isDrink={true}
      loading={false}
    />
  </View>
</Layout>
`;
